Compilation failure on basic example [np_pythran]

Not sure if this is a Pythran issue, or where I should report it if not, but I get a compiler error on the following Cython code:
```
# cython: np_pythran=True
# cython: cxx=True

import numpy as np
cimport numpy as np

def test():
    cdef np.ndarray[double, ndim=2] A = np.random.rand(1, 1)
    cdef double a
    with nogil:
        a = A[0, 0]
```
The compiler error is impressively long, and culminates in:
```
/home/terhorst/opt/py37/lib/python3.7/site-packages/pythran/pythonic/include/types/numpy_vexpr.hpp:78:31: error: request for member ‘fast’ in ‘((const {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int, long int> >, long int>*)this)->{anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int, long int> >, long int>::view_’, which is of non-class type ‘
const long int’
   78 |       return data_.fast(view_.fast(i));
      |                         ~~~~~~^~~~
```

--- a/pythran/pythonic/include/types/ndarray.hpp
+++ b/pythran/pythonic/include/types/ndarray.hpp
@@ -433,6 +433,12 @@ namespace types
         extended_slice<count_new_axis<S0, S...>::value>{}((*this), s0, s...));
 
     template <class S0, class... S>
+        auto operator()(S0 const &s0, S const &... s) &
+        -> decltype(extended_slice<count_new_axis<S0, S...>::value>{}((*this),
+                                                                      s0,
+                                                                      s...));
+
+    template <class S0, class... S>
         auto operator()(S0 const &s0, S const &... s) &&
         -> decltype(extended_slice<count_new_axis<S0, S...>::value>{}(
             std::move(*this), s0, s...));
@@ -550,19 +556,19 @@ namespace types
                                           : contiguous_slice())...);
     }
 
-    template <class Ty0, class Ty1, class... Tys>
+    template <class Ty0, class Ty1, class... Tys,
+              class _ = typename std::enable_if<is_numexpr_arg<Ty0>::value,
+                                                void>::type>
     auto operator[](std::tuple<Ty0, Ty1, Tys...> const &indices) const ->
         typename std::enable_if<is_numexpr_arg<Ty0>::value,
                                 decltype(this->_fwdindex(
                                     indices, utils::make_index_sequence<
                                                  2 + sizeof...(Tys)>()))>::type;
 
-    template <class Ty, size_t M>
-    auto operator[](array<Ty, M> const &indices) const & ->
-        typename std::enable_if<
-            !std::is_integral<Ty>::value,
-            decltype(this->_fwdindex(indices,
-                                     utils::make_index_sequence<M>()))>::type
+    template <class Ty, size_t M, class _ = typename std::enable_if<
+                                      !std::is_integral<Ty>::value, void>::type>
+    auto operator[](array<Ty, M> const &indices) const
+        & -> decltype(this->_fwdindex(indices, utils::make_index_sequence<M>()))
     {
       return _fwdindex(indices, utils::make_index_sequence<M>());
     }

--- a/pythran/pythonic/include/types/numpy_gexpr.hpp
+++ b/pythran/pythonic/include/types/numpy_gexpr.hpp
@@ -69,32 +69,17 @@ namespace types
    */
   template <size_t C>
   struct extended_slice {
-    template <class T, class pS, class... S>
-    auto operator()(ndarray<T, pS> &&a, S const &... s)
-        -> decltype(std::move(a).reshape(make_reshape<C>(
-            a.shape(),
+    template <class E, class... S>
+    auto operator()(E &&expr, S const &... s)
+        -> decltype(std::forward<E>(expr).reshape(make_reshape<C>(
+            expr.shape(),
             std::tuple<
                 std::integral_constant<bool, to_slice<S>::is_new_axis>...>()))(
             to_slice<S>{}(s)...))
 
     {
-      return std::move(a).reshape(make_reshape<C>(
-          a.shape(),
-          std::tuple<
-              std::integral_constant<bool, to_slice<S>::is_new_axis>...>()))(
-          to_slice<S>{}(s)...);
-    }
-
-    template <class T, class pS, class... S>
-    auto operator()(ndarray<T, pS> const &a, S const &... s)
-        -> decltype(a.reshape(make_reshape<C>(
-            a.shape(),
-            std::tuple<
-                std::integral_constant<bool, to_slice<S>::is_new_axis>...>()))(
-            to_slice<S>{}(s)...))
-    {
-      return a.reshape(make_reshape<C>(
-          a.shape(),
+      return std::forward<E>(expr).reshape(make_reshape<C>(
+          expr.shape(),
           std::tuple<
               std::integral_constant<bool, to_slice<S>::is_new_axis>...>()))(
           to_slice<S>{}(s)...);
@@ -103,75 +88,57 @@ namespace types
 
   template <>
   struct extended_slice<0> {
-    template <class T, class pS, class... S>
-    auto operator()(ndarray<T, pS> &&a, long const &s0, S const &... s)
-        -> decltype(std::declval<numpy_iexpr<ndarray<T, pS>>>()(s...))
+    template <class E, class... S>
+    auto operator()(E &&expr, long const &s0, S const &... s) ->
+        typename std::enable_if<
+            utils::all_of<std::is_integral<S>::value...>::value,
+            decltype(std::forward<E>(expr)[types::make_tuple(s0, s...)])>::type
     {
-      return std::move(a)[s0](s...);
+      return std::forward<E>(expr)[types::make_tuple(s0, s...)];
+    }
+    template <class E, class... S>
+    auto operator()(E &&expr, long const &s0, S const &... s) ->
+        typename std::enable_if<
+            !utils::all_of<std::is_integral<S>::value...>::value,
+            decltype(std::forward<E>(expr)[s0](s...))>::type
+    {
+      return std::forward<E>(expr)[s0](s...);
     }
 
-    template <class T, class pS, class... S>
-    auto operator()(ndarray<T, pS> const &a, long const &s0, S const &... s)
-        -> decltype(a[s0](s...))
+    template <class E, class... S, size_t... Is>
+    numpy_gexpr<typename std::decay<E>::type, normalize_t<S>...>
+    fwd(E &&expr, std::tuple<S...> const &s, utils::index_sequence<Is...>)
     {
-      return a[s0](s...);
+      return {std::forward<E>(expr),
+              std::get<Is>(s).normalize(std::get<Is>(expr.shape()))...};
     }
 
-    template <class T, class pS, class... S, size_t... Is>
-    numpy_gexpr<ndarray<T, pS>, normalize_t<S>...>
-    fwd(ndarray<T, pS> &&a, std::tuple<S...> const &s,
-        utils::index_sequence<Is...>)
-    {
-      return {std::move(a),
-              std::get<Is>(s).normalize(std::get<Is>(a.shape()))...};
-    }
-
-    template <class T, class pS, class... S, size_t... Is>
-    numpy_gexpr<ndarray<T, pS>, normalize_t<S>...>
-    fwd(ndarray<T, pS> const &a, std::tuple<S...> const &s,
-        utils::index_sequence<Is...>)
-    {
-      return {a, std::get<Is>(s).normalize(std::get<Is>(a.shape()))...};
-    }
-
-    template <class T, class pS, class... S>
-    numpy_gexpr<ndarray<T, pS>, normalized_slice, normalize_t<S>...>
-    operator()(ndarray<T, pS> &&a, slice const &s0, S const &... s)
-    {
-      return fwd(std::move(a), std::make_tuple(s0, s...),
-                 utils::make_index_sequence<sizeof...(S) + 1>());
-    }
-
-    template <class T, class pS, class... S>
-    numpy_gexpr<ndarray<T, pS> const &, normalized_slice, normalize_t<S>...>
-    operator()(ndarray<T, pS> const &a, slice const &s0, S const &... s)
-    {
-      return fwd(a, std::make_tuple(s0, s...),
-                 utils::make_index_sequence<sizeof...(S) + 1>());
-    }
-
-    template <class T, class pS, class... S>
-    numpy_gexpr<ndarray<T, pS>, contiguous_normalized_slice, normalize_t<S>...>
-    operator()(ndarray<T, pS> &&a, contiguous_slice const &s0, S const &... s)
-    {
-      return make_gexpr(std::move(a), s0, s...);
-    }
-
-    template <class T, class pS, class... S>
-    numpy_gexpr<ndarray<T, pS> const &, contiguous_normalized_slice,
+    template <class E, class... S>
+    numpy_gexpr<typename std::decay<E>::type, normalized_slice,
                 normalize_t<S>...>
-    operator()(ndarray<T, pS> const &a, contiguous_slice const &s0,
-               S const &... s)
+    operator()(E &&expr, slice const &s0, S const &... s)
     {
-      return make_gexpr(a, s0, s...);
+      return fwd(std::forward<E>(expr), std::make_tuple(s0, s...),
+                 utils::make_index_sequence<sizeof...(S) + 1>());
     }
 
-    template <class T, class pS, class F, class... S>
-    numpy_gexpr<ndarray<T, array<long, std::tuple_size<pS>::value>>,
-                contiguous_normalized_slice, normalize_t<S>...>
-    operator()(ndarray<T, pS> const &a, F const &s0, S const &... s)
+    template <class E, class... S>
+    numpy_gexpr<typename std::decay<E>::type, contiguous_normalized_slice,
+                normalize_t<S>...>
+    operator()(E &&expr, contiguous_slice const &s0, S const &... s)
     {
-      return numpy_vexpr<ndarray<T, pS>, F>{a, s0}(
+      return make_gexpr(std::forward<E>(expr), s0, s...);
+    }
+
+    template <class E, class F, class... S>
+    numpy_gexpr<ndarray<typename std::decay<E>::type::dtype,
+                        array<long, std::decay<E>::type::value>>,
+                contiguous_normalized_slice, normalize_t<S>...>
+    operator()(E &&expr, F const &s0, S const &... s)
+    {
+      return numpy_vexpr<ndarray<typename std::decay<E>::type::dtype,
+                                 array<long, std::decay<E>::type::value>>,
+                         F>{std::forward<E>(expr), s0}(
           contiguous_slice(none_type{}, none_type{}), s...);
     }
   };

--- a/pythran/pythonic/types/ndarray.hpp
+++ b/pythran/pythonic/types/ndarray.hpp
@@ -759,6 +759,15 @@ namespace types
 
   template <class T, class pS>
       template <class S0, class... S>
+      auto ndarray<T, pS>::operator()(S0 const &s0, S const &... s) &
+      -> decltype(extended_slice<count_new_axis<S0, S...>::value>{}((*this), s0,
+                                                                    s...))
+  {
+    return extended_slice<count_new_axis<S0, S...>::value>{}((*this), s0, s...);
+  }
+
+  template <class T, class pS>
+      template <class S0, class... S>
       auto ndarray<T, pS>::operator()(S0 const &s0, S const &... s) &&
       -> decltype(extended_slice<count_new_axis<S0, S...>::value>{}(
           std::move(*this), s0, s...))
@@ -851,7 +860,7 @@ namespace types
   }
 
   template <class T, class pS>
-  template <class Ty0, class Ty1, class... Tys>
+  template <class Ty0, class Ty1, class... Tys, class _>
   auto ndarray<T, pS>::
   operator[](std::tuple<Ty0, Ty1, Tys...> const &indices) const ->
       typename std::enable_if<

--- a/pythran/tests/cython/indexing.pyx
+++ b/pythran/tests/cython/indexing.pyx
@@ -1,0 +1,13 @@
+# cython: np_pythran=True
+
+import numpy as np
+cimport numpy as np
+
+def indexing():
+    cdef np.ndarray[double, ndim=2] A = np.random.rand(1, 1)
+    cdef double a
+    with nogil:
+        A[0, 0] = 12.
+        a = A[0, 0]
+    return a
+

--- a/pythran/tests/cython/setup_indexing.py
+++ b/pythran/tests/cython/setup_indexing.py
@@ -1,0 +1,13 @@
+from distutils.core import setup
+from Cython.Build import cythonize
+
+setup(
+    name = "indexing",
+    ext_modules = cythonize('indexing.pyx'),
+    script_name = 'setup.py',
+    script_args = ['build_ext', '--inplace']
+)
+import indexing
+import numpy as np
+print(indexing.indexing())
+

